3-05-project-view-header

IevgeniiaAbdulina opened the pull request:
1. Task: [RS Clone – Angular Final Project](https://github.com/rolling-scopes-school/tasks/blob/master/angular/modules/rsclone/README.md)
2. Screenshot:
<img width="682" height="224" alt="project-header-1" src="https://github.com/user-attachments/assets/901f5c56-ad2f-49ba-8f02-d7b798c3970c" />
<img width="683" height="258" alt="project-header-2" src="https://github.com/user-attachments/assets/8e8e9b86-086e-4b67-9777-d8d08feae732" />
<img width="475" height="384" alt="project-header-4" src="https://github.com/user-attachments/assets/d59c6e17-86bd-4d1e-a525-71633e100e0f" />
<img width="475" height="332" alt="project-header-3" src="https://github.com/user-attachments/assets/ed57969f-706a-4bd0-966e-76d8834f4ab6" />

3. Deploy: [FoxFlow Task Management app](https://foxflowangular2025.web.app/main)
5. Done: 2025-09-23 / Deadline: 2025-10-06

## The feature list:

- [x] - Project header element
  - [x]  - Project icon (optional)
  - [x]  - Project name
  - [x]  - Actions icon button (Dropdown arrow for small screen)
  - [x] - Edit Project title.
  - [x] - Delete Project.
  - [ ]  - Project owner icon (will be done after updating the Registration service and saving the users in Firebase)

IevgeniiaAbdulina committed:
feat: Integrate Project Data with Tasks Data.

ALIAKSEI-sl commented on src/app/features/components/kanban-board/kanban-board.ts:
```diff
@@ -9,4 +19,34 @@ import { TranslateModule } from '@ngx-translate/core';
+      });
+      this.allTasks.set(newTasks);
+
+      this.toDoTasks.set(tasks.filter((task) => task.status === 'todo'));
+      this.inProgressTasks.set(
+        tasks.filter((task) => task.status === 'inprogress')
+      );
+      this.doneTasks.set(tasks.filter((task) => task.status === 'done'));
```
you do the same logic three times, use cycle

IevgeniiaAbdulina replied:
In my opinion, the circle is not an optimized option. We need to update the signal whenever we set a task item with a particular condition. But in the case of a filter, we set a signal item only once.

ALIAKSEI-sl replied:
You don't need to set one task every time during the cycle. However you are supposed to group tasks within one cycle and then update signals.
You can use this https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/groupBy or reduce

IevgeniiaAbdulina replied:
Is it correct solution?

```ts
const groupedTasks = tasks.reduce(
  (acc, task) => {
    if (!acc[task.status]) {
      acc[task.status] = [];
    }
    acc[task.status].push(task);
    return acc;
  },
  {} as Record<string, TaskData[]>
);

this.toDoTasks.set(groupedTasks['todo'] || []);
this.inProgressTasks.set(groupedTasks['inprogress'] || []);
this.doneTasks.set(groupedTasks['done'] || []);
```

ALIAKSEI-sl commented on src/app/features/components/kanban-board/kanban-board.ts:
```diff
@@ -9,4 +19,34 @@ import { TranslateModule } from '@ngx-translate/core';
+    const projectId = this.projectService.currentProject().id as string;
+    console.log('Hello project', projectId);
+    this.tasksService.getTasksFromFirebase(projectId).subscribe((tasks) => {
+      // for every task which does not have status set 'todo'
```
Why can task not to have status? Must be status by default at list.

IevgeniiaAbdulina replied:
This functionality was created for tasks created earlier that did not yet have a status. It will be removed in future updates.

IevgeniiaAbdulina replied:
removed

ALIAKSEI-sl commented on src/app/features/components/kanban-board/kanban-board.ts:
```diff
@@ -9,4 +19,34 @@ import { TranslateModule } from '@ngx-translate/core';
+        }
+        return task;
+      });
+      this.allTasks.set(newTasks);
```
Do you need all tasks?

IevgeniiaAbdulina replied:
Unused allTasks object has been removed.

ALIAKSEI-sl commented on src/app/features/services/projects-service/project-service.ts:
```diff
@@ -0,0 +1,26 @@
+    title: '',
+    owner: '',
+    createdAt: '',
+  });
```
Could  you use null here?

ALIAKSEI-sl commented on src/app/features/services/projects-service/projects-firebase-service.ts:
```diff
@@ -49,7 +51,10 @@ export class ProjectsFirebaseService {
     const promise = addDoc(this.projectsCollection, projectToCreate).then(
-      (result) => result.id
+      (result) => {
+        this.projectService.addTasksCollection(result.id);
```
Why you create task after adding project?

IevgeniiaAbdulina replied:
To demonstrate the task possibilities, a default first task is shown.

ALIAKSEI-sl replied:
Does it still need?

IevgeniiaAbdulina replied:
removed

ALIAKSEI-sl commented on src/app/services/firebase/firebase-service.ts:
```diff
@@ -40,7 +32,12 @@ export class FirebaseServiceTs {
-    const todoToCreate = { title: text, isCompleted: false, createdAt: date };
+    const todoToCreate = {
+      title: text,
+      isCompleted: false,
+      createdAt: date,
+      status: 'todo',
```
There are too much arguments, pass an object.

IevgeniiaAbdulina replied:
What did you mean here? What do you suggest?

ALIAKSEI-sl replied:
If function has more than three arguments, it'd be better pass one object.

ALIAKSEI-sl commented on src/app/shared/components/task-list-item/task-list-item.ts:
```diff
@@ -22,31 +22,35 @@ import { CommonModule } from '@angular/common';
-    this.editingText = this.task().title;
+    this.taskId = this.task()?.id ?? '';
+    this.editingText = this.task()?.title ?? '';
+    this.projectId = this.projectService.currentProject().id as string;
```
Could you use computed here?

IevgeniiaAbdulina replied:
Replaced with:
`  readonly projectId = computed(() => this.projectService.currentProject()?.id);`
- the other part of the logic will be refactored after merging with Olga's part of code (Tasks service)

ALIAKSEI-sl commented on src/app/shared/components/projects-list-item/projects-list-item.ts:
```diff
@@ -20,8 +21,10 @@ import { Router } from '@angular/router';
+  private projectService = inject(ProjectService);
 
   openProject(): void {
+    this.projectService.currentProject.set({ ...this.project$() });
```
If you don't display any information about project you don't have to save, you can take projectId from url

IevgeniiaAbdulina replied:
Removed unused code.

IevgeniiaAbdulina committed:
refactor: update routing, add Project header, update projectId computed.

IevgeniiaAbdulina committed:
refactor: grouped tasks by status.

IevgeniiaAbdulina committed:
refactor: grouped tasks by status and update the latest ts config.

IevgeniiaAbdulina committed:
refactor: update empty condition.

IevgeniiaAbdulina committed:
feat: Add rename an delete Project.

IevgeniiaAbdulina committed:
feat: Add Responsive Project header, use breakpointObserver.

IevgeniiaAbdulina committed:
refactor: update project title font size.

IevgeniiaAbdulina committed:
refactor: removed unused code.

IevgeniiaAbdulina committed:
refactor: removed unused imports.

IevgeniiaAbdulina committed:
feat: Add custom password validator.

ALIAKSEI-sl commented on src/app/auth/components/register/register.ts:
```diff
@@ -73,10 +73,30 @@ export class Register {
+    },
+  ];
+
+  hasError(errorKey: string): boolean {
+    const errors = this.form.get('password')?.errors;
+    return !!errors && errors[errorKey];
+  }
```
move this logic to Directive

ALIAKSEI-sl commented on src/app/auth/components/register/register.ts:
```diff
@@ -73,10 +73,30 @@ export class Register {
-      [Validators.required, Validators.pattern(StrongPasswordRegx)],
+      [
+        Validators.required,
+        Validators.pattern(StrongPasswordRegx),
```
What does this validator do?

ALIAKSEI-sl commented on src/app/features/components/home-page/home-page.ts:
```diff
@@ -30,20 +32,26 @@ import { take } from 'rxjs';
+        this.projects.set(result);
+      });
+  }
+
```
And what have been changed?

ALIAKSEI-sl commented on src/app/features/components/home-page/home-page.ts:
```diff
@@ -30,20 +32,26 @@ import { take } from 'rxjs';
+  ngOnInit(): void {
+    this.projectsFirebaseService
+      .getProjects()
+      .subscribe((result: Project[]) => {
```
Could you unsubscribe?

ALIAKSEI-sl commented on src/app/features/components/kanban-board/kanban-board.ts:
```diff
@@ -9,4 +19,33 @@ import { TranslateModule } from '@ngx-translate/core';
+  projectId: string | undefined;
+
+  ngOnInit(): void {
+    this.route.params.subscribe((params) => {
```
Could you create a separate method?

ALIAKSEI-sl commented on src/app/features/components/kanban-board/kanban-board.ts:
```diff
@@ -9,4 +19,33 @@ import { TranslateModule } from '@ngx-translate/core';
+      if (this.projectId) {
+        this.tasksService
+          .getTasksFromFirebase(this.projectId)
+          .subscribe((tasks) => {
```
subscription inside subscription? use swithMap und unsubscribe

ALIAKSEI-sl commented on src/app/features/components/project-header/project-header.html:
```diff
@@ -0,0 +1,96 @@
+    </div>
+  </mat-label>
+  } @else {
+  <ng-container>
+    <mat-form-field
+      appearance="outline"
+      class="project-form-field">
+      <mat-label>Project title</mat-label>
+      <input
+        #titleInput
+        matInput
+        type="text"
+        [(ngModel)]="editingText"
+        (keydown.enter)="saveEditing()"
+        (keydown.escape)="cancelEditing()" />
+    </mat-form-field>
+  </ng-container>
```
What is the purpose of ng-container here?

ALIAKSEI-sl commented on src/app/features/components/project-header/project-header.ts:
```diff
@@ -0,0 +1,141 @@
+    this.setResponsiveHeader();
+  }
+
+  private setResponsiveHeader(): void {
+    this.breakpointObserver
+      .observe(['(max-width: 800px)'])
+      .pipe(takeUntilDestroyed(this.destroyRef))
+      .subscribe((screenSize) => {
+        if (screenSize.matches) {
+          this.isShowActions.set(false);
+        } else {
+          this.isShowActions.set(true);
+        }
+      });
+  }
```
аж слов не хватает, если уже и подписываться, тогда просто this.isShowActions.set(!screenSize.matches);
но в чем логика, подписаться на observable, чтобы просто засетать это в сигнал, а потом использовать в темплейте, вместо того, чтобы сделать это через AsyncPipe, он так и просится здесь, просто нужно добавить map.

ALIAKSEI-sl commented on src/app/features/services/projects-service/project-service.ts:
```diff
@@ -0,0 +1,51 @@
+    this.projectsFirebaseService
+      .updateProject(editingId, dataToUpdate)
+      .subscribe(() => {
+        this.getProject(editingId);
```
Are you sure you neeв call getProject here? It doesnt look like web socket.

ALIAKSEI-sl commented on src/app/shared/interfaces/task-interface.ts:
```diff
@@ -1,7 +1,7 @@
   id: string;
   title: string;
-  //status: 'todo' | 'inprogress' | 'done';
+  status: 'todo' | 'inprogress' | 'done';
```
type?

ALIAKSEI-sl commented on src/app/shared/utils/password-validator.ts:
```diff
@@ -0,0 +1,27 @@
+import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';
+
+export function passwordValidator(): ValidatorFn {
+  return (control: AbstractControl): ValidationErrors | null => {
+    const password: string = control.value;
+
+    const errors: ValidationErrors = [];
+
+    if (!/(?=.*[A-Z])/.test(password)) {
+      errors['uppercase'] = 'At least one uppercase letter.';
+    }
+    if (!/(?=.*[a-z])/.test(password)) {
+      errors['lowercase'] = 'At least one lowercase letter.';
+    }
+    if (!/(?=.*[0-9])/.test(password)) {
+      errors['digit'] = 'At least one digit.';
+    }
+    if (!/(?=.*[!@#$%^&*])/.test(password)) {
+      errors['specialChar'] = 'At least one special character (!@#$%^&*).';
+    }
+    if (password.length < 8) {
+      errors['min_length'] = 'At least 8 characters long.';
+    }
+
+    return Object.keys(errors).length ? errors : null;
+  };
+}
```
что это за старнный JS? 
type ValidationErrors = {
  [key: string]: any;
}
почему errors это массив? 
у массива ключи должны .snm только числа , что это например 
errors['uppercase'] = 'At least one uppercase letter.'; ????

IevgeniiaAbdulina committed:
refactor: update after rebase.
> 1. Remove Tasks Service.
> 2. Update Tasks name collection.
> 3. Remove unused code and comments.
> 4. Update imports.

rekunolya reviewed (approved):
Looks good! Hope you fix remarks on next branch)
